internal/store/generic.go

```diff
@@ -0,0 +1,399 @@
+		return false, flterrors.ErrResourceVersionConflict
+	}
+
+	resource.SetResourceVersion(lo.ToPtr(lo.FromPtr(existing.GetResourceVersion()) + 1))
+
+	selectFields := []string{"spec"}
+	if resource.GetKind() == api.DeviceKind {
+		selectFields = append(selectFields, "alias")
+	}
+	selectFields = append(selectFields, s.getNonNilFieldsFromResource(resource)...)
+	selectFields = append(selectFields, fieldsToUnset...)
+
+	query := s.db.WithContext(ctx).Model(resource).
+		Where("org_id = ? AND name = ? AND (resource_version IS NULL OR resource_version = ?)",
+			resource.GetOrgID(),
+			resource.GetName(),
+			lo.FromPtr(existing.GetResourceVersion())).
+		Select(selectFields)
+
+	result := query.Updates(resource)
+	if result.Error != nil {
+		return false, ErrorFromGormError(result.Error)
+	}
+	if result.RowsAffected == 0 {
+		return true, flterrors.ErrNoRowsUpdated
+	}
+	return false, nil
+}
+
+func (s *GenericStore[P, M, A, ML, AL]) getNonNilFieldsFromResource(resource P) []string {
+	ret := []string{}
+	if resource.GetGeneration() != nil {
+		ret = append(ret, "generation")
+	}
+	if resource.GetLabels() != nil {
+		ret = append(ret, "labels")
+	}
+	if resource.GetOwner() != nil {
+		ret = append(ret, "owner")
+	}
+	if resource.GetAnnotations() != nil {
+		ret = append(ret, "annotations")
+	}
+	if resource.GetResourceVersion() != nil {
+		ret = append(ret, "resource_version")
+	}
+
+	return ret
+}
+
+func (s *GenericStore[P, M, A, ML, AL]) Get(ctx context.Context, orgId uuid.UUID, name string) (*A, error) {
+	resource, err := s.getModel(ctx, orgId, name)
+	if err != nil {
+		return nil, err
+	}
+	apiResource, err := s.modelPtrToAPI(s.modelPtrToModel(resource))
+	return apiResource, err
+}
+
+func (s *GenericStore[P, M, A, ML, AL]) getModel(ctx context.Context, orgId uuid.UUID, name string) (*M, error) {
+	var resource M
+	result := s.db.WithContext(ctx).Where("org_id = ? AND name = ? AND spec IS NOT NULL", orgId, name).First(&resource)
+	if result.Error != nil {
+		return nil, ErrorFromGormError(result.Error)
+	}
+
+	return &resource, nil
+}
+
+func (s *GenericStore[P, M, A, ML, AL]) Delete(ctx context.Context, resource M, callback func(before, after *M), associatedResources ...Resource) error {
+	if len(associatedResources) == 0 {
+		return s.delete(ctx, resource, callback)
+	}
+	return s.deleteWithAssociated(ctx, resource, callback, associatedResources...)
+}
+
+func (s *GenericStore[P, M, A, ML, AL]) delete(ctx context.Context, resource M, callback func(before, after *M)) error {
+	result := s.db.WithContext(ctx).Unscoped().Where("spec IS NOT NULL").Delete(&resource)
+	if result.Error != nil {
+		return ErrorFromGormError(result.Error)
+	}
+	if result.RowsAffected == 0 {
+		return nil
+	}
+
+	if callback != nil {
+		callback(&resource, nil)
+	}
+	return nil
+}
+
+func (s *GenericStore[P, M, A, ML, AL]) deleteWithAssociated(ctx context.Context, resource M, callback func(before, after *M), associatedResources ...Resource) error {
```

Comment:
Once this has converged, it'd be great if you added docs comments to these methods.

Reply:
Will do